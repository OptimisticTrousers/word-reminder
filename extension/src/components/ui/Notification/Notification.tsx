import { CircleCheck, CircleX, X } from "lucide-react";
import { useContext } from "react";
import CSSModules from "react-css-modules";

import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import styles from "./Notification.module.css";

export const Notification = CSSModules(
  function ({ type, message }: { type: string; message: string }) {
    const { dismissNotification } = useContext(NotificationContext);

    function handleClick() {
      dismissNotification();
    }

    return (
      <div
        styleName={`notification notification--${type}`}
        role="dialog"
        aria-labelledby="notification-message"
      >
        <div styleName="notification__flex">
          <div styleName="notification__item">
            {type === NOTIFICATION_ACTIONS.SUCCESS ? (
              <CircleCheck
                data-testid="check-icon"
                styleName="notification__icon notification__icon--circle"
              />
            ) : (
              <CircleX
                data-testid="x-icon"
                styleName="notification__icon notification__icon--circle"
              />
            )}
          </div>
          <div styleName="notification__box">
            <span id="notification-message" styleName="notification__message">
              {message}
            </span>
          </div>
          <div styleName="notification__controls">
            <div styleName="notification__control">
              <button styleName="notification_button" onClick={handleClick}>
                <span styleName="notification__text">Dismiss</span>
                <X styleName="notification__icon" />
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
