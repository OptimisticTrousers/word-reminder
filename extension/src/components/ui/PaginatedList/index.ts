export { PaginatedList } from "./PaginatedList";
export type { Props } from "./PaginatedList";
