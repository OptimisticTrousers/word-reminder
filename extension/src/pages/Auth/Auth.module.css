.auth {
  padding: 1rem;
}

.auth__container {
  margin-bottom: 1rem;
}

.auth__form {
  margin-top: 0.75rem;
}

.auth__image {
  width: 250px;
  height: 250px;
}

.auth__navigation {
  display: flex;
  justify-content: space-between;
  align-items: center;
  box-shadow: 1px 3px 3px rgba(0, 0, 0, 0.35);
  margin-bottom: 1rem;
  border: 1px solid var(--border-color);
  padding: 0.75rem;
  border-radius: 12p;
}

.auth__control {
  margin-bottom: 0.75rem;
}

.auth__label {
  color: var(--label-text);
  font-size: 0.75rem;
  margin-bottom: 0.5rem;
}

.auth__input {
  margin-top: 0.25rem;
  width: 100%;
  color: var(--label-text-color);
}

.auth__input:disabled {
  filter: grayscale(0.5) opacity(0.5);
  pointer-events: none;
  -webkit-filter: grayscale(0.5) opacity(0.5);
}

.auth__button {
  color: var(--button-text-color);
}

.auth__button--login {
  margin-bottom: 0.75rem;
}

.auth__link {
  color: var(--secondary-anchor-text-color);
}

.auth__link--create {
  margin-left: 0.25rem;
}
