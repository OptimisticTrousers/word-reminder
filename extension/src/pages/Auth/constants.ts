export const AUTH_NOTIFICATION_MSGS = {
  credentialsExpired: () => {
    return "Your credentials have expired. Please log in again.";
  },
};
