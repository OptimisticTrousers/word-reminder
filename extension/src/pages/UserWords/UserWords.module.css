.words__top,
.words__form {
  margin-bottom: 0.6rem;
}

.words__control {
  flex: 1 1 auto;
}

.words__or {
  align-content: center;
}

.words__form {
  display: flex;
  justify-content: space-between;
  align-items: center;
  gap: 0.75rem;
  flex-wrap: wrap;
}

.words__control,
.words__label {
  display: flex;
  align-items: center;
}

.words__button,
.words__input {
  flex: 1;
}

.words__button {
  display: flex;
  align-items: center;
  justify-content: center;
}

.sort__select,
.words__input,
.words__button--import,
.words__select,
.words__icon {
  margin-left: 0.5rem;
}

.words__icon--export {
  margin-right: 0.5rem;
}
