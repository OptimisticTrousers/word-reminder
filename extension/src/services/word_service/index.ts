export { wordService } from "./word_service";
