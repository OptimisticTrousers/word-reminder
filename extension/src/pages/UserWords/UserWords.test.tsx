import { WORD_MAX } from "common";
import { createRoutesStub, Outlet } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { AUTH_NOTIFICATION_MSGS } from "../Auth/constants";
import { Props } from "../../components/ui/PaginatedList";
import { NotificationProvider } from "../../context/Notification";
import { wordService } from "../../services/word_service/word_service";
import { UserWords } from "./UserWords";
import * as utils from "../../utils/download";

vi.mock("../../components/ui/PaginatedList", function () {
  return {
    PaginatedList: function (props: Props) {
      return (
        <>
          <div data-testid="name">{props.name}</div>
          <div data-testid="totalRows">{props.totalRows}</div>
          <div data-testid="list">{props.list}</div>
          <div data-testid="isLoading">{props.isLoading.toString()}</div>
          <div data-testid="isSuccess">{props.isSuccess.toString()}</div>
          <div data-testid="error">{JSON.stringify(props.error)}</div>
          <div data-testid="previous">{JSON.stringify(props.previous)}</div>
          <div data-testid="next">{JSON.stringify(props.next)}</div>
        </>
      );
    },
  };
});

vi.mock("../../components/words/UserWord", function () {
  return {
    UserWord: function (props: Props) {
      return (
        <>
          <div data-testid="props">{JSON.stringify(props)}</div>
        </>
      );
    },
  };
});

describe("UserWords component", () => {
  const json = [
    {
      id: "1",
      word: "word",
    },
  ];

  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

  function setup() {
    const path = "/words";
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path,
            Component: UserWords,
          },
          {
            path: "/login",
            Component: function () {
              return <form></form>;
            },
          },
        ],
      },
    ]);

    return {
      user: userEvent.setup(),
      ...render(
        <NotificationProvider>
          <QueryClientProvider client={queryClient}>
            <Stub initialEntries={[path]} />
          </QueryClientProvider>
        </NotificationProvider>
      ),
    };
  }

  const testUser = {
    id: "1",
  };

  const PAGINATION_LIMIT = "10";

  beforeEach(() => {
    vi.clearAllMocks();
    vi.spyOn(wordService, "getWordList").mockImplementation(async () => {
      return {
        json: {
          userWords: json,
          totalRows: 1,
          previous: {
            page: 1,
            limit: PAGINATION_LIMIT,
          },
          next: { page: 3, limit: PAGINATION_LIMIT },
        },
        status: 200,
      };
    });
  });

  describe("query words", () => {
    describe("paginated list", () => {
      it("error response", async () => {
        const message = "Bad Request.";
        const status = 400;
        vi.spyOn(wordService, "getWordList").mockImplementation(async () => {
          return Promise.reject({ json: { message }, status });
        });

        setup();

        await waitFor(() => {
          const name = screen.getByTestId("name");
          const list = screen.getByTestId("list");
          const totalRows = screen.getByTestId("totalRows");
          const isLoading = screen.getByTestId("isLoading");
          const isSuccess = screen.getByTestId("isSuccess");
          const error = screen.getByTestId("error");
          const previous = screen.getByTestId("previous");
          const next = screen.getByTestId("next");
          expect(name).toHaveTextContent("Words");
          expect(list).toBeEmptyDOMElement();
          expect(totalRows).toBeEmptyDOMElement();
          expect(isLoading).toHaveTextContent("false");
          expect(isSuccess).toHaveTextContent("false");
          expect(error).toHaveTextContent(JSON.stringify({ message }));
          expect(previous).toHaveTextContent("");
          expect(next).toHaveTextContent("");
        });
      });

      it("loading response", async () => {
        const status = 200;
        const delay = 500;
        vi.spyOn(wordService, "getWordList").mockImplementation(async () => {
          return new Promise((resolve) => {
            setTimeout(() => {
              resolve({
                json: {
                  userWords: json,
                  totalRows: 1,
                  previous: {
                    page: 1,
                    limit: PAGINATION_LIMIT,
                  },
                  next: { page: 3, limit: PAGINATION_LIMIT },
                },
                status,
              });
            }, delay);
          });
        });

        setup();

        await waitFor(() => {
          const name = screen.getByTestId("name");
          const list = screen.getByTestId("list");
          const totalRows = screen.getByTestId("totalRows");
          const isLoading = screen.getByTestId("isLoading");
          const isSuccess = screen.getByTestId("isSuccess");
          const error = screen.getByTestId("error");
          const previous = screen.getByTestId("previous");
          const next = screen.getByTestId("next");
          expect(name).toHaveTextContent("Words");
          expect(list).toBeEmptyDOMElement();
          expect(totalRows).toBeEmptyDOMElement();
          expect(isLoading).toHaveTextContent("true");
          expect(isSuccess).toHaveTextContent("false");
          expect(error).toHaveTextContent(JSON.stringify(null));
          expect(previous).toHaveTextContent("");
          expect(next).toHaveTextContent("");
        });
      });

      it("success response", async () => {
        const status = 200;
        vi.spyOn(wordService, "getWordList").mockImplementation(async () => {
          return {
            json: {
              userWords: json,
              totalRows: 1,
              previous: {
                page: 1,
                limit: PAGINATION_LIMIT,
              },
              next: { page: 3, limit: PAGINATION_LIMIT },
            },
            status,
          };
        });

        setup();

        await waitFor(() => {
          const name = screen.getByTestId("name");
          const list = screen.getByTestId("list");
          const totalRows = screen.getByTestId("totalRows");
          const isLoading = screen.getByTestId("isLoading");
          const isSuccess = screen.getByTestId("isSuccess");
          const error = screen.getByTestId("error");
          const previous = screen.getByTestId("previous");
          const next = screen.getByTestId("next");
          expect(name).toHaveTextContent("Words");
          expect(list).not.toBeEmptyDOMElement();
          expect(totalRows).toHaveTextContent("1");
          expect(isLoading).toHaveTextContent("false");
          expect(isSuccess).toHaveTextContent("true");
          expect(error).toHaveTextContent(JSON.stringify(null));
          expect(previous).toHaveTextContent(
            JSON.stringify({ page: 1, limit: PAGINATION_LIMIT })
          );
          expect(next).toHaveTextContent(
            JSON.stringify({ page: 3, limit: PAGINATION_LIMIT })
          );
        });
      });
    });

    describe("searching", () => {
      it("calls the functions to make a search query", async () => {
        const { user } = setup();
        const status = 200;
        const mockWordServiceGetUserWords = vi
          .spyOn(wordService, "getWordList")
          .mockImplementation(async () => {
            return { json: { userWords: json, totalRows: 1 }, status };
          });

        const searchInput = screen.getByLabelText("Search", {
          selector: "input",
        });
        const search = "word";
        await user.type(searchInput, search);
        const filterButton = screen.getByRole("button", { name: "Filter" });
        await user.click(filterButton);

        expect(mockWordServiceGetUserWords).toHaveBeenCalledTimes(1);
        expect(mockWordServiceGetUserWords).toHaveBeenCalledWith(testUser.id, {
          page: "1",
          limit: PAGINATION_LIMIT,
          search,
          learned: "",
          column: "",
          direction: "",
        });
      });
    });

    describe("sorting", () => {
      it("does not call the functions to filter by featured because it is the default option", async () => {
        const { user } = setup();
        const status = 200;
        const mockWordServiceGetUserWords = vi
          .spyOn(wordService, "getWordList")
          .mockImplementation(async () => {
            return { json: { userWords: json, totalRows: 1 }, status };
          });

        const sortSelect = screen.getByRole("combobox", { name: "Sort by:" });
        await user.selectOptions(sortSelect, ["Featured"]);
        const filterButton = screen.getByRole("button", { name: "Filter" });
        await user.click(filterButton);

        expect(mockWordServiceGetUserWords).not.toHaveBeenCalled();
      });

      it("calls the functions to sort by newest based on 'created_at' field", async () => {
        const { user } = setup();
        const status = 200;
        const mockWordServiceGetUserWords = vi
          .spyOn(wordService, "getWordList")
          .mockImplementation(async () => {
            return { json: { userWords: json, totalRows: 1 }, status };
          });

        const sortSelect = screen.getByRole("combobox", { name: "Sort by:" });
        await user.selectOptions(sortSelect, ["Newest"]);
        const filterButton = screen.getByRole("button", { name: "Filter" });
        await user.click(filterButton);

        expect(mockWordServiceGetUserWords).toHaveBeenCalledTimes(1);
        expect(mockWordServiceGetUserWords).toHaveBeenCalledWith(testUser.id, {
          page: "1",
          limit: PAGINATION_LIMIT,
          search: "",
          learned: "",
          column: "created_at",
          direction: "1",
        });
      });

      it("calls the functions to sort by oldest based on 'created_at' field", async () => {
        const { user } = setup();
        const status = 200;
        const mockWordServiceGetUserWords = vi
          .spyOn(wordService, "getWordList")
          .mockImplementation(async () => {
            return { json: { userWords: json, totalRows: 1 }, status };
          });

        const sortSelect = screen.getByRole("combobox", { name: "Sort by:" });
        await user.selectOptions(sortSelect, ["Oldest"]);
        const filterButton = screen.getByRole("button", { name: "Filter" });
        await user.click(filterButton);

        expect(mockWordServiceGetUserWords).toHaveBeenCalledTimes(1);
        expect(mockWordServiceGetUserWords).toHaveBeenCalledWith(testUser.id, {
          page: "1",
          limit: PAGINATION_LIMIT,
          search: "",
          learned: "",
          column: "created_at",
          direction: "-1",
        });
      });
    });

    describe("filtering", () => {
      it("does calls the functions to sort based on any words because it is the default", async () => {
        const { user } = setup();
        const status = 200;
        const mockWordServiceGetUserWords = vi
          .spyOn(wordService, "getWordList")
          .mockImplementation(async () => {
            return { json: { userWords: json, totalRows: 1 }, status };
          });

        const filterSelect = screen.getByRole("combobox", {
          name: "Filter by:",
        });
        await user.selectOptions(filterSelect, ["Any"]);
        const filterButton = screen.getByRole("button", { name: "Filter" });
        await user.click(filterButton);

        expect(mockWordServiceGetUserWords).not.toHaveBeenCalled();
      });

      it("calls the functions to sort based on learned words", async () => {
        const { user } = setup();
        const status = 200;
        const mockWordServiceGetUserWords = vi
          .spyOn(wordService, "getWordList")
          .mockImplementation(async () => {
            return { json: { userWords: json, totalRows: 1 }, status };
          });

        const filterSelect = screen.getByRole("combobox", {
          name: "Filter by:",
        });
        await user.selectOptions(filterSelect, ["Learned"]);
        const filterButton = screen.getByRole("button", { name: "Filter" });
        await user.click(filterButton);

        expect(mockWordServiceGetUserWords).toHaveBeenCalledTimes(1);
        expect(mockWordServiceGetUserWords).toHaveBeenCalledWith(testUser.id, {
          page: "1",
          limit: PAGINATION_LIMIT,
          search: "",
          learned: "true",
          column: "",
          direction: "",
        });
      });

      it("calls the functions to sort based on unlearned words", async () => {
        const { user } = setup();
        const status = 200;
        const mockWordServiceGetUserWords = vi
          .spyOn(wordService, "getWordList")
          .mockImplementation(async () => {
            return { json: { userWords: json, totalRows: 1 }, status };
          });

        const filterSelect = screen.getByRole("combobox", {
          name: "Filter by:",
        });
        await user.selectOptions(filterSelect, ["Unlearned"]);
        const filterButton = screen.getByRole("button", { name: "Filter" });
        await user.click(filterButton);

        expect(mockWordServiceGetUserWords).toHaveBeenCalledTimes(1);
        expect(mockWordServiceGetUserWords).toHaveBeenCalledWith(testUser.id, {
          page: "1",
          limit: PAGINATION_LIMIT,
          search: "",
          learned: "false",
          column: "",
          direction: "",
        });
      });
    });
  });

  describe("word creation", () => {
    describe("text word creation", () => {
      it("calls the functions to create a word", async () => {
        const formData = new FormData();
        const word = json[0].word;
        formData.append("word", word);
        formData.append("csv", new File([""], ""));
        formData.append("userId", testUser.id);
        const status = 200;
        const mockWordServiceCreateWord = vi
          .spyOn(wordService, "createWord")
          .mockImplementation(async () => {
            return { json: { word: json[0] }, status };
          });
        const mockInvalidateQueries = vi.spyOn(
          queryClient,
          "invalidateQueries"
        );
        const { user } = setup();

        const wordInput = screen.getByLabelText("Word", {
          selector: "input",
        });
        await user.type(wordInput, word);
        const addButton = screen.getByRole("button", {
          name: "Add",
        });
        await user.click(addButton);

        const notification = screen.getByRole("dialog", {
          name: "You have successfully added a word to your dictionary.",
        });
        expect(mockInvalidateQueries).toHaveBeenCalledWith({
          queryKey: [
            "words",
            Object.fromEntries(
              new URLSearchParams({
                page: "1",
                limit: PAGINATION_LIMIT,
                search: "",
                learned: "",
                column: "",
                direction: "",
              })
            ),
          ],
          exact: true,
        });
        expect(mockWordServiceCreateWord).toHaveBeenCalledTimes(1);
        expect(mockWordServiceCreateWord).toHaveBeenCalledWith({
          userId: testUser.id,
          formData,
        });
        expect(notification).toBeInTheDocument();
      });

      it("shows generic notification that there was an error", async () => {
        const formData = new FormData();
        const word = json[0].word;
        formData.append("word", word);
        formData.append("csv", new File([""], ""));
        formData.append("userId", testUser.id);
        const message = "Bad Request.";
        const status = 400;
        const mockWordServiceCreateWord = vi
          .spyOn(wordService, "createWord")
          .mockImplementation(async () => {
            return Promise.reject({ json: { message }, status });
          });
        const { user } = setup();

        const wordInput = screen.getByLabelText("Word", {
          selector: "input",
        });
        await user.type(wordInput, word);
        const addButton = screen.getByRole("button", {
          name: "Add",
        });
        await user.click(addButton);

        const notification = screen.getByRole("dialog", {
          name: message,
        });
        expect(mockWordServiceCreateWord).toHaveBeenCalledTimes(1);
        expect(mockWordServiceCreateWord).toHaveBeenCalledWith({
          userId: testUser.id,
          formData,
        });
        expect(notification).toBeInTheDocument();
      });

      it("shows notification notification that the user is unauthenticated", async () => {
        const formData = new FormData();
        const word = json[0].word;
        formData.append("word", word);
        formData.append("csv", new File([""], ""));
        formData.append("userId", testUser.id);
        const message = "Unauthenticated.";
        const status = 401;
        const mockWordServiceCreateWord = vi
          .spyOn(wordService, "createWord")
          .mockImplementation(async () => {
            return Promise.reject({ json: { message }, status });
          });
        const { user } = setup();

        const wordInput = screen.getByLabelText("Word", {
          selector: "input",
        });
        await user.type(wordInput, word);
        const addButton = screen.getByRole("button", {
          name: "Add",
        });
        await user.click(addButton);

        const notification = screen.getByRole("dialog", {
          name: AUTH_NOTIFICATION_MSGS.credentialsExpired(),
        });
        expect(mockWordServiceCreateWord).toHaveBeenCalledTimes(1);
        expect(mockWordServiceCreateWord).toHaveBeenCalledWith({
          userId: testUser.id,
          formData,
        });
        expect(notification).toBeInTheDocument();
      });

      it("calls the functions at the word limit to create a word when the word is too long", async () => {
        const word = new Array(WORD_MAX + 1).fill("a").join("");
        const formData = new FormData();
        formData.append("word", word.slice(0, -1));
        formData.append("csv", new File([""], ""));
        formData.append("userId", testUser.id);
        const status = 200;
        const mockWordServiceCreate = vi
          .spyOn(wordService, "createWord")
          .mockImplementation(async () => {
            return { json: { word: json[0] }, status };
          });
        const { user } = setup();

        const wordInput = screen.getByLabelText("Word", {
          selector: "input",
        });
        await user.type(wordInput, word);
        const addButton = screen.getByRole("button", {
          name: "Add",
        });
        await user.click(addButton);

        const notification = screen.getByRole("dialog", {
          name: "You have successfully added a word to your dictionary.",
        });
        expect(mockWordServiceCreate).toHaveBeenCalledTimes(1);
        expect(mockWordServiceCreate).toHaveBeenCalledWith({
          userId: testUser.id,
          formData,
        });
        expect(notification).toBeInTheDocument();
      });
    });

    describe("csv", () => {
      describe("export words", () => {
        it("calls the functions to download csv file", async () => {
          const { user } = setup();
          const totalRows = "1";
          const status = 200;
          const mockWordServiceGetUserWords = vi
            .spyOn(wordService, "getWordList")
            .mockImplementation(async () => {
              return {
                json: {
                  userWords: [{ word: "export" }, { word: "word" }],
                  totalRows,
                },
                status,
              };
            });
          const mockDownload = vi
            .spyOn(utils, "download")
            .mockImplementation(vi.fn());

          const exportWordsButton = screen.getByRole("button", {
            name: "Export Words",
          });
          await user.click(exportWordsButton);

          expect(mockWordServiceGetUserWords).toHaveBeenCalledTimes(1);
          expect(mockWordServiceGetUserWords).toHaveBeenCalledWith(
            testUser.id,
            {
              page: "1",
              limit: totalRows,
              search: "",
              learned: "",
              column: "",
              direction: "",
            }
          );
          expect(mockDownload).toHaveBeenCalledTimes(1);
          expect(mockDownload).toHaveBeenCalledWith({
            data: "export,word",
            fileName: "words.csv",
            fileType: "text/csv",
          });
        });
      });

      describe("import words", async () => {
        it("calls the functions to import words from a csv file", async () => {
          const formData = new FormData();
          const csvFile = new File(["many,words,in,this,file"], "words.csv", {
            type: "text/csv",
          });
          formData.append("word", "");
          formData.append("csv", csvFile);
          formData.append("userId", testUser.id);
          const { user } = setup();
          const status = 200;
          const mockWordServiceCreate = vi
            .spyOn(wordService, "createWord")
            .mockImplementation(async () => {
              return { json: { message: "1 word has been created." }, status };
            });

          const importWordsInput: HTMLInputElement = screen.getByLabelText(
            "Import Words",
            { selector: "input" }
          );
          await user.upload(importWordsInput, csvFile);
          const addButton = screen.getByRole("button", { name: "Add" });
          await user.click(addButton);

          const notification = screen.getByRole("dialog", {
            name: "You have successfully multiple words to your dictionary.",
          });
          expect(notification).toBeInTheDocument();
          expect(mockWordServiceCreate).toHaveBeenCalledTimes(1);
          expect(mockWordServiceCreate).toHaveBeenCalledWith({
            userId: testUser.id,
            formData,
          });
          expect(importWordsInput.files?.[0]).toBe(csvFile);
          expect(importWordsInput.files?.item(0)).toBe(csvFile);
          expect(importWordsInput.files).toHaveLength(1);
        });

        it("does not call the functions to import words when it is not a csv file", async () => {
          const formData = new FormData();
          const notACsvFile = new File(["(⌐□_□)"], "words.png", {
            type: "image/png",
          });
          formData.append("word", "");
          formData.append("csv", notACsvFile);
          formData.append("userId", testUser.id);
          const status = 200;
          const mockWordServiceCreate = vi
            .spyOn(wordService, "createWord")
            .mockImplementation(async () => {
              return { json, status };
            });
          const { user } = setup();

          const importWordsInput: HTMLInputElement = screen.getByLabelText(
            "Import Words",
            {
              selector: "input",
            }
          );
          await user.upload(importWordsInput, notACsvFile);
          const addButton = screen.getByRole("button", {
            name: "Add",
          });
          await user.click(addButton);

          const notification = screen.getByRole("dialog", {
            name: "Please type a word or upload a CSV file of words.",
          });
          expect(mockWordServiceCreate).not.toHaveBeenCalled();
          expect(notification).toBeInTheDocument();
          expect(importWordsInput.files).toHaveLength(0);
        });

        it("shows generic notification that there was an error", async () => {
          const formData = new FormData();
          const csvFile = new File(["many,words,in,this,file"], "words.csv", {
            type: "text/csv",
          });
          formData.append("word", "");
          formData.append("csv", csvFile);
          formData.append("userId", testUser.id);
          const message = "Bad Request.";
          const status = 400;
          const mockWordServiceCreateWord = vi
            .spyOn(wordService, "createWord")
            .mockImplementation(async () => {
              return Promise.reject({ json: { message }, status });
            });
          const { user } = setup();

          const importWordsInput: HTMLInputElement = screen.getByLabelText(
            "Import Words",
            {
              selector: "input",
            }
          );
          await user.upload(importWordsInput, csvFile);
          const addButton = screen.getByRole("button", {
            name: "Add",
          });
          await user.click(addButton);

          const notification = screen.getByRole("dialog", {
            name: message,
          });
          expect(mockWordServiceCreateWord).toHaveBeenCalledTimes(1);
          expect(mockWordServiceCreateWord).toHaveBeenCalledWith({
            userId: testUser.id,
            formData,
          });
          expect(notification).toBeInTheDocument();
          expect(importWordsInput.files?.[0]).toBe(csvFile);
          expect(importWordsInput.files?.item(0)).toBe(csvFile);
          expect(importWordsInput.files).toHaveLength(1);
        });

        it("shows notification notification that the user is unauthenticated", async () => {
          const formData = new FormData();
          const csvFile = new File(["many,words,in,this,file"], "words.csv", {
            type: "text/csv",
          });
          formData.append("word", "");
          formData.append("csv", csvFile);
          formData.append("userId", testUser.id);
          const message = "Unauthenticated.";
          const status = 401;
          const mockWordServiceCreateWord = vi
            .spyOn(wordService, "createWord")
            .mockImplementation(async () => {
              return Promise.reject({ json: { message }, status });
            });
          const { user } = setup();

          const importWordsInput: HTMLInputElement = screen.getByLabelText(
            "Import Words",
            {
              selector: "input",
            }
          );
          await user.upload(importWordsInput, csvFile);
          const addButton = screen.getByRole("button", {
            name: "Add",
          });
          await user.click(addButton);

          const notification = screen.getByRole("dialog", {
            name: AUTH_NOTIFICATION_MSGS.credentialsExpired(),
          });
          expect(mockWordServiceCreateWord).toHaveBeenCalledTimes(1);
          expect(mockWordServiceCreateWord).toHaveBeenCalledWith({
            userId: testUser.id,
            formData,
          });
          expect(notification).toBeInTheDocument();
          expect(importWordsInput.files?.[0]).toBe(csvFile);
          expect(importWordsInput.files?.item(0)).toBe(csvFile);
          expect(importWordsInput.files).toHaveLength(1);
        });
      });
    });

    it("shows a notification when neither a word or file is provided", async () => {
      const formData = new FormData();
      const word = "";
      formData.append("word", word);
      formData.append("csv", new File([""], ""));
      formData.append("userId", testUser.id);
      const status = 200;
      const mockWordServiceCreateWord = vi
        .spyOn(wordService, "createWord")
        .mockImplementation(async () => {
          return { json: { word: json[0] }, status };
        });
      const { user } = setup();

      const addButton = screen.getByRole("button", {
        name: "Add",
      });
      await user.click(addButton);

      const notification = screen.getByRole("dialog", {
        name: "Please type a word or upload a CSV file of words.",
      });
      expect(mockWordServiceCreateWord).not.toHaveBeenCalled();
      expect(notification).toBeInTheDocument();
    });

    it("disables add button and all inputs when the mutation is pending", async () => {
      const formData = new FormData();
      const word = json[0].word;
      formData.append("word", word);
      formData.append("csv", new File([""], ""));
      formData.append("userId", testUser.id);
      const delay = 500;
      const status = 200;
      const mockWordServiceCreateWord = vi
        .spyOn(wordService, "createWord")
        .mockImplementation(async () => {
          return new Promise((resolve) => {
            setTimeout(() => {
              resolve({ json: { word: json[0] }, status });
            }, delay);
          });
        });
      const { user } = setup();

      const wordInput = screen.getByLabelText("Word", {
        selector: "input",
      });
      await user.type(wordInput, word);
      const addButton = screen.getByRole("button", {
        name: "Add",
      });
      await user.click(addButton);

      expect(mockWordServiceCreateWord).toHaveBeenCalledTimes(1);
      expect(mockWordServiceCreateWord).toHaveBeenCalledWith({
        userId: testUser.id,
        formData,
      });
      await waitFor(() => {
        const addButton = screen.getByRole("button", { name: "Add" });
        const inputs = screen.getAllByLabelText(/[\s\S]*/, {
          selector: "input",
        });
        expect(addButton).toBeDisabled();
        for (const input of inputs) {
          expect(input).toBeDisabled();
        }
      });
    });
  });

  it("renders two forms that have fields for creating a text word, importing a csv file of words, searching words, sorting words, and filtering words", async () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });
});
