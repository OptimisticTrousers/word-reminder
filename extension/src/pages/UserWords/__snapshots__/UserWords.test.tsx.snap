// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserWords component > renders two forms that have fields for creating a text word, importing a csv file of words, searching words, sorting words, and filtering words 1`] = `
<DocumentFragment>
  <section
    class="_words_727679"
  >
    <div
      class="_words__top_727679"
    >
      <div
        class="_words__container_727679"
      >
        <form
          action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
          class="_words__form_727679"
        >
          <div
            class="_words__control_727679"
          >
            <label
              class="_words__label_727679"
            >
              Word
              <input
                class="_words__input_727679 _words__input--add_727679"
                maxlength="45"
                name="word"
                placeholder="Add a new word to your dictionary..."
                type="text"
              />
            </label>
          </div>
          <div
            class="_words__control_727679"
          >
            <label
              class="_words__label_727679"
            >
              Import Words
              <svg
                class="lucide lucide-import _word__icon_727679"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 3v12"
                />
                <path
                  d="m8 11 4 4 4-4"
                />
                <path
                  d="M8 5H4a2 2 0 0 0-2 2v10a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V7a2 2 0 0 0-2-2h-4"
                />
              </svg>
              <input
                accept="text/csv"
                class="_words__button_727679 _words__button--import_727679"
                name="csv"
                type="file"
              />
            </label>
          </div>
          <button
            class="_words__button_727679 _words__button--add_727679"
            type="submit"
          >
            Add
          </button>
        </form>
        <form
          action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
          class="_words__form_727679"
        >
          <div
            class="_words__control_727679"
          >
            <label
              class="_words__label_727679"
            >
              Export Words
              <svg
                class="lucide lucide-download _word__icon_727679"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
                />
                <polyline
                  points="7 10 12 15 17 10"
                />
                <line
                  x1="12"
                  x2="12"
                  y1="15"
                  y2="3"
                />
              </svg>
              <input
                class="_words__button_727679 _words__button--import_727679"
                type="button"
              />
            </label>
          </div>
          <div
            class="_words__control_727679"
          >
            <label
              class="_words__label_727679"
            >
              Search
              <input
                class="_words__input_727679 _words__input--search_727679"
                name="search"
                placeholder="Search for words in your dictionary..."
                type="search"
              />
            </label>
          </div>
          <div
            class="_words__control_727679"
          >
            <label
              class="_words__label_727679"
            >
              Sort by:
              <select
                class="_words__select_727679"
                name="created_at"
              >
                <option
                  class="_words__option_727679"
                  value=""
                >
                  Featured
                </option>
                <option
                  class="_words__option_727679"
                  value="1"
                >
                  Newest
                </option>
                <option
                  class="_words__option_727679"
                  value="-1"
                >
                  Oldest
                </option>
              </select>
            </label>
          </div>
          <div
            class="_words__control_727679"
          >
            <label
              class="_words__label_727679"
            >
              Filter by:
              <select
                class="_words__select_727679"
                name="learned"
              >
                <option
                  class="_words__option_727679"
                  value=""
                >
                  Any
                </option>
                <option
                  class="_words__option_727679"
                  value="true"
                >
                  Learned
                </option>
                <option
                  class="_words__option_727679"
                  value="false"
                >
                  Unlearned
                </option>
              </select>
            </label>
          </div>
          <button
            class="_words__button_727679 _words__button--add_727679"
            type="submit"
          >
            Filter
          </button>
        </form>
      </div>
    </div>
    <div
      data-testid="name"
    >
      Words
    </div>
    <div
      data-testid="totalRows"
    >
      1
    </div>
    <div
      data-testid="list"
    >
      <div
        data-testid="props"
      >
        {"id":"1","word":"word"}
      </div>
    </div>
    <div
      data-testid="isLoading"
    >
      false
    </div>
    <div
      data-testid="isSuccess"
    >
      true
    </div>
    <div
      data-testid="error"
    >
      null
    </div>
    <div
      data-testid="previous"
    >
      {"page":1,"limit":"10"}
    </div>
    <div
      data-testid="next"
    >
      {"page":3,"limit":"10"}
    </div>
  </section>
</DocumentFragment>
`;
